src/components/auth/recover/HasRecoveryKit.svelte

```diff
@@ -0,0 +1,152 @@
+<script lang="ts">
+  import * as RootKey from 'webnative/common/root-key'
+  import * as UCAN from 'webnative/ucan/index'
+  import * as DID from 'webnative/did/index'
+  import * as uint8arrays from 'uint8arrays'
+
+  import { sessionStore } from '$src/stores'
+  import {
+    RECOVERY_STATES,
+    USERNAME_STORAGE_KEY,
+    loadAccount,
+    prepareUsername
+  } from '$lib/auth/account'
+  import Check from '$components/icons/CheckIcon.svelte'
+  import RecoveryKitButton from './RecoveryKitButton.svelte'
+
+  let state = $sessionStore.session
+    ? RECOVERY_STATES.Done
+    : RECOVERY_STATES.Ready
+
+  /**
+   * Parse the user's `username` and `readKey` from the uploaded recovery kit and pass them into
+   * webnative to recover the user's account and populate the `session` and `filesystem` stores
+   * @param files
+   */
+  export const handleFileInput: (
+    files: FileList
+  ) => Promise<void> = async files => {
+    const reader = new FileReader()
+
+    reader.onload = async event => {
+      state = RECOVERY_STATES.Processing
+
+      try {
+        const {
+          authStrategy,
+          program: {
+            components: { crypto, reference, storage }
+          }
+        } = $sessionStore
+
+        const parts = (event.target.result as string)
+          .split('username: ')[1]
+          .split('key: ')
+
+        const readKey = uint8arrays.fromString(
+          parts[1].replace(/(\r\n|\n|\r)/gm, ''),
+          'base64pad'
+        )
+
+        const oldUsername = parts[0].replace(/(\r\n|\n|\r)/gm, '')
+        const hashedOldUsername = await prepareUsername(oldUsername)
+        const oldRootDID = await reference.didRoot.lookup(hashedOldUsername)
+
+        // Construct a new username using the old `trimmed` name and `oldRootDID`
+        const newUsername = `${oldUsername.split('#')[0]}#${oldRootDID}`
+        const hashedNewUsername = await prepareUsername(newUsername)
+
+        storage.setItem(USERNAME_STORAGE_KEY, newUsername)
+
+        // Register the user with the `hashedNewUsername`
+        const { success } = await authStrategy.register({
+          username: hashedNewUsername
+        })
+        if (!success) {
+          throw new Error('Failed to register new user')
+        }
+
+        // Build an ephemeral UCAN to allow the
+        const issuer = await DID.write(crypto)
```

Comment:
You can also say `program.agentDID()` here, whatever you prefer.

Reply:
done 👍🏼